opening_hours/models.py

```diff
@@ -0,0 +1,103 @@
+
+    def __str__(self) -> str:
+        return str(self.id)
+
+    def is_reservable(self, start_datetime: datetime, end_datetime: datetime) -> bool:
+        return self.reservable_time_spans.filter_period(start=start_datetime, end=end_datetime).exists()
+
+
+class ReservableTimeSpanQuerySet(models.QuerySet):
+    def filter_period(self, start: datetime | date, end: datetime | date) -> QuerySet["ReservableTimeSpan"]:
+        """Filter reservable time spans that overlap with the given period."""
+        # Convert dates to datetimes to include timezone information
+        if isinstance(start, date):
+            start = datetime.combine(start, time.min, tzinfo=DEFAULT_TIMEZONE)
+        if isinstance(end, date):
+            end = datetime.combine(end, time.min, tzinfo=DEFAULT_TIMEZONE) + timedelta(days=1)
+
+        return self.filter(start_datetime__lt=end, end_datetime__gt=start)
+
+    def filter_date(self, selected_date: datetime | date) -> QuerySet["ReservableTimeSpan"]:
+        """Filter reservable time spans that overlap with the given date."""
+        return self.filter_period(start=selected_date, end=selected_date)
+
+    def truncated_start_and_end_datetimes_for_period(self, start: date, end: date) -> QuerySet["ReservableTimeSpan"]:
+        """
+        Annotate truncated start and end datetimes for reservable time spans that overlap with the given period.
+
+        If the time span starts before the period, the start time is set to the period start.
+        If the time span ends after the period, the end time is set to the period end.
+        """
+        start_datetime = datetime.combine(start, time.min, tzinfo=DEFAULT_TIMEZONE)
+        end_datetime = datetime.combine(end, time.max, tzinfo=DEFAULT_TIMEZONE) + timedelta(days=1)
+        return self.filter_period(start=start, end=end).annotate(
+            truncated_start_datetime=Case(
+                When(
+                    condition=Q(start_datetime__lt=start_datetime),
+                    then=Value(start_datetime),
+                ),
+                default="start_datetime",
+            ),
+            truncated_end_datetime=Case(
+                When(
+                    condition=Q(end_datetime__gt=end_datetime),
+                    then=Value(end_datetime),
+                ),
+                default="end_datetime",
+            ),
+        )
+
+
+class ReservableTimeSpan(models.Model):
+    """A time period on which a ReservationUnit is reservable."""
+
+    resource = models.ForeignKey(
+        OriginHaukiResource,
+        related_name="reservable_time_spans",
+        on_delete=models.CASCADE,
+    )
+    start_datetime = models.DateTimeField(null=False, blank=False)
+    end_datetime = models.DateTimeField(null=False, blank=False)
+
+    objects = ReservableTimeSpanQuerySet.as_manager()
+
+    class Meta:
+        ordering = ["resource", "start_datetime", "end_datetime"]
+        constraints = [
+            models.CheckConstraint(
```

Comment:
How about a unique constraint that there can't be a duplicate timespan for the same resource?

Reply:
In theory there should never be a duplicate timespan for a resource, and even if there was, i dont think that other related functionality would be negatively affected by those :D

Reply:
I guess given the current functionality duplicates don't really affect things. 

But, what about later when the frontend fetches reservable timespans for the calendar view on the reservation page? If there is a duplicate, would that affect drawing that view?

On the other hand, adding the constraint would mean that you'd need to modify the `ReservableTimeSpanClient` to account for duplicates, possibly with a hashmap instead of a list somewhere in gathering the TimeSpanElements. 

I'd add the constraint in favor of data correctness, but I'll leave it to you to what you want to do.

Reply:
The `ReservableTimeSpanClient` already checks if the last-already-existing `ReservableTimeSpan` overlaps with the first-new-fetched-time-span, so duplicates should already be handled.

Also in `_init_date_range` fetching past dates is prevented by limiting the `start_date`, so we shouldn't be able to fetch any dates before our last existing `ReservableTimeSpan`.

Reply:
Now that I think about it more, I guess the whole point of `ReservableTimeSpanClient` is to prevent duplicate ReservableTimeSpan elements due to the merging/splitting rules. So, I think it would be safe to add the constraint without any modifications.